src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+                       'HLS playlist. Switching to another playlist.');
+
+      return this.masterPlaylistLoader_.media(nextPlaylist);
+    }
+    videojs.log.warn('Problem encountered with the current ' +
+                     'HLS playlist. No suitable alternatives found.');
+    // We have no more playlists we can select so we must fail
+    this.hlsHandler.error = error;
+    return this.mediaSource.endOfStream('network');
+  }
+
+  pause() {
+    this.mainSegmentLoader_.pause();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.pause();
+    }
+  }
+
+  setCurrentTime(currentTime) {
+    let buffered = Ranges.findRange_(this.hlsHandler.tech_.buffered(), currentTime);
+
+    if (!(this.masterPlaylistLoader_ && this.masterPlaylistLoader_.media())) {
+      // return immediately if the metadata is not ready yet
+      return 0;
+    }
+
+    // it's clearly an edge-case but don't thrown an error if asked to
+    // seek within an empty playlist
+    if (!this.masterPlaylistLoader_.media().segments) {
+      return 0;
+    }
+
+    // if the seek location is already buffered, continue buffering as
+    // usual
+    if (buffered && buffered.length) {
+      return currentTime;
+    }
+
+    // cancel outstanding requests so we begin buffering at the new
+    // location
+    this.mainSegmentLoader_.abort();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.abort();
+    }
+
+    if (!this.hlsHandler.tech_.paused()) {
+      this.mainSegmentLoader_.load();
+      if (this.audioPlaylistLoader_) {
+        this.audioSegmentLoader_.load();
+      }
+    }
+  }
+
+  duration() {
+    if (!this.masterPlaylistLoader_) {
+      return 0;
+    }
+
+    if (this.mediaSource) {
+      return this.mediaSource.duration;
+    }
+
+    return Hls.Playlist.duration(this.masterPlaylistLoader_.media());
+  }
+
+  seekable() {
+    let media;
+    let seekable;
+
+    if (!this.masterPlaylistLoader_) {
+      return videojs.createTimeRanges();
+    }
+    media = this.masterPlaylistLoader_.media();
+    if (!media) {
+      return videojs.createTimeRanges();
+    }
+
+    seekable = Hls.Playlist.seekable(media);
+    if (seekable.length === 0) {
+      return seekable;
+    }
+
+    // if the seekable start is zero, it may be because the player has
+    // been paused for a long time and stopped buffering. in that case,
+    // fall back to the playlist loader's running estimate of expired
+    // time
+    if (seekable.start(0) === 0) {
+      return videojs.createTimeRanges([[this.masterPlaylistLoader_.expired_,
+                                        this.masterPlaylistLoader_.expired_ +
+                                          seekable.end(0)]]);
+    }
+
+    // seekable has been calculated based on buffering video data so it
+    // can be returned directly
+    return seekable;
+  }
+
+  /**
+   * Update the player duration
+   */
+  updateDuration(playlist) {
```

Comment:
In putting the segment loader together, I came to the conclusion that we're doing this totally wrong :) I think we should set the duration when the first media playlist loads and then let media sources manage it from that point out. As it stands, I think we may be invoking the coded frame removal algorithm unnecessarily.